cpp/src/fil/infer.cu

```diff
@@ -177,92 +212,178 @@ struct tree_aggregator_t {
   }
 
   __device__ __forceinline__ void finalize(float* out, int num_rows,
-                                           int output_stride) {
+                                           int output_stride,
+                                           output_t transform, int num_trees) {
     __syncthreads();
-    typedef typename BlockReduce<NITEMS>::TempStorage TempStorage;
-    acc = BlockReduce<NITEMS>(*(TempStorage*)tmp_storage).Sum(acc);
+    acc = block_reduce(acc, vectorized(cub::Sum()), tmp_storage);
     if (threadIdx.x > 0) return;
 #pragma unroll
-    for (int row = 0; row < num_rows; ++row)
-      out[row * output_stride] = acc[row];
+    for (int row = 0; row < NITEMS; ++row)
+      if(row < num_rows) out[row * output_stride] = acc[row];
```

Comment:
as @canonizer requested, making sure the unrolled loop has a compile-time-defined number of iterations